src/OrchardCore.Modules/OrchardCore.Queries/Sql/SqlQuerySource.cs

```diff
@@ -71,7 +72,20 @@ public async Task<IQueryResults> ExecuteQueryAsync(Query query, IDictionary<stri
                 IEnumerable<long> documentIds;
 
                 using var transaction = await connection.BeginTransactionAsync(_session.Store.Configuration.IsolationLevel);
-                documentIds = await connection.QueryAsync<long>(rawQuery, parameters, transaction);
+                var queryResult = await connection.QueryAsync(rawQuery, parameters, transaction);
+
+                documentIds = queryResult.Select(row =>
+                {
+                    var rowDict = (IDictionary<string, object>)row;
+
+                    if (rowDict.TryGetValue(nameof(ContentItemIndex.DocumentId), out var docIdObj) &&
+                        docIdObj is long docId)
+                    {
+                        return docId;
+                    }
+
+                    return rowDict.FirstOrDefault(kv => kv.Value is long).Value as long? ?? 0;
```

Comment:
I think this is the same as before. But I didn't check `QueryAsync<long>` in detail.

Reply:
not really. You are still looking for a DocumentId before that logic. On the other side, maybe all we need it to throw an exception if no column called `DocumentId` was found so we are not guessing. The improvment here if someone query using `SELECT * FROM`, `SELECT DocumentId FROM` or other like `SELECT Id, DocumentId FROM` we'll return the correct result. But if domeone query `SELECT Id FROM` they would get an exception.

Reply:
Before, `QueryAsync<long>` took the first column that has a `long` value. That's how I fell into this trap. 

Doing it my way will also allow e.g. `SELECT DocumentId as Foo FROM` etc. We can't solve every case, as you pointed out. But it's more likely to give correct answers now.

Reply:
> I think this is the same as before. But I didn't check `QueryAsync<long>` in detail.

Oh, I think this was misunderstood. I meant before the whole change, not your idea.

Reply:
Yes I get that. And even worst is that if you use Select * from would likely to return the Id as long not DocumentId as long as which would likely to return no documents.

The issue I see with the proposed code is that you are working with dictionary for each row. So in one row you are looking for the first value in the row that can be cases from object to long. This means in some rows the first value maybe return by in other rows 2 value "from a different column" maybe returned. So it would not be consistent. Which is why I think as a fallback evaluate only the first value as we need in the past just so that old behavior do not change but at least we are looking for DocumentId column here.

Reply:
That is correct, but the dictionary is the internal representation anyway. So I guess `QeryAsync<long>` was doing basically the same before, maybe a little bit more sophisticated. 
We can of cause keep the column consistent. I originally just wanted a quick solution, but I understand your concern :)

Reply:
What about something like this:

```csharp
string column = null;
var columnSelected = false;
documentIds = queryResult.Select(row =>
{
    var rowDictionary = (IDictionary<string, object>)row;

    if(!columnSelected)
    {
        columnSelected = true;
        if (rowDictionary.ContainsKey(nameof(ContentItemIndex.DocumentId)))
        {
            column = nameof(ContentItemIndex.DocumentId);
        }
        else
        {
            column = rowDictionary.FirstOrDefault(kv => kv.Value is long).Key;
        }
    }

    return column != null && rowDictionary.TryGetValue(column, out var documentIdObject) && documentIdObject is long documentId
        ? documentId
        : 0;
});
```

Reply:
I think `QueryAsync<long>` uses data reader to fetch each row and then always takes the first column and attempt to cast it. Always **first column** not checking for the **first matching** value of the given type.

Reply:
That is good. I would just tweak it a little like this instead

```

string column = null;

documentIds = queryResult.Select(row =>
{
    var rowDictionary = (IDictionary<string, object>)row;

    if (column == null)
    {
        if (rowDictionary.ContainsKey(nameof(ContentItemIndex.DocumentId)))
        {
            column = nameof(ContentItemIndex.DocumentId);
        }
        else
        {
            column = rowDictionary.FirstOrDefault(kv => kv.Value is long).Key 
            ?? rowDictionary.First().Key;
        }
    }

    return rowDictionary.TryGetValue(column, out var documentIdObject) && documentIdObject is long documentId
        ? documentId
        : 0;
});
```